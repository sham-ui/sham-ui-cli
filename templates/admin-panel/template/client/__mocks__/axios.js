const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};
const INTERCEPTORS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const request = {
                url,
                method,
                headers: {}
            };
            INTERCEPTORS.request( request );
            const { status, data, headers } = FIXTURES[ method ][ url ];
            if ( 200 === status ) {
                const response = {
                    data,
                    headers,
                    status
                };
                return Promise.resolve(
                    INTERCEPTORS.response.success( response )
                );
            } else {
                const error = {
                    config: {
                        url,
                        method,
                        baseURL: '/'
                    },
                    response: {
                        data,
                        headers,
                        status
                    }
                };
                return INTERCEPTORS.response.fail( error );
            }
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);

const defaultMocksData = {
    csrfToken: 'csrf-token',
    user: {
        Name: 'John Smith',
        Email: 'j.smith@example.com'
    }
};

export default {
    defaultMocksData: defaultMocksData,
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        const instance  = {
            ...MOCKS,
            request( { url, method = 'get', data } ) {
                const args = [ url ];
                if ( undefined !== data ) {
                    args.push( data );
                }
                return instance[ method ].apply( null, args );
            },
            interceptors: {
                request: {
                    use( interceptor ) {
                        INTERCEPTORS.request = interceptor;
                    }
                },
                response: {
                    use( success, fail ) {
                        INTERCEPTORS.response = {
                            success,
                            fail
                        };
                    }
                }
            }
        };
        return instance;
    } ),
    use( method, url, data, status = 200, headers = {} ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"` );
        }
        FIXTURES[ method ][ url ] = {
            data,
            status,
            headers
        };
        return this;
    },

    useDefaultMocks() {
        this
            .use( 'get', '/csrftoken', '', 200, { 'x-csrf-token': defaultMocksData.csrfToken } )
            .use( 'get', '/validsession', defaultMocksData.user );
        return this;
    }
};
