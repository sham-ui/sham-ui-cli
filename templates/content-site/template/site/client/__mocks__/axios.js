const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};
const INTERCEPTORS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const request = {
                url,
                method,
                headers: {}
            };
            if ( INTERCEPTORS.request ) {
                INTERCEPTORS.request( request );
            }
            const { status, data, headers } = FIXTURES[ method ][ url ];
            if ( 200 === status ) {
                const response = {
                    data,
                    headers,
                    status
                };
                return Promise.resolve(
                    INTERCEPTORS.response.success( response )
                );
            } else {
                const error = {
                    config: {
                        url,
                        method,
                        baseURL: '/'
                    },
                    response: {
                        data,
                        headers,
                        status
                    }
                };
                return INTERCEPTORS.response.fail( error );
            }
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);


export default {
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        const instance  = {
            ...MOCKS,
            request( { url, method = 'get', data, params } ) {
                const args = [ url ];
                if ( undefined !== data ) {
                    args.push( data );
                }
                if ( undefined !== params ) {
                    args.push( params );
                }
                return instance[ method ].apply( null, args );
            },
            interceptors: {
                request: {
                    use( interceptor ) {
                        INTERCEPTORS.request = interceptor;
                    }
                },
                response: {
                    use( success, fail ) {
                        INTERCEPTORS.response = {
                            success,
                            fail
                        };
                    }
                }
            }
        };
        return instance;
    } ),
    use( method, url, data, status = 200, headers = {} ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"` );
        }
        FIXTURES[ method ][ url ] = {
            data,
            status,
            headers
        };
        return this;
    },

    useDefaultMocks() {
        return this;
    }
};
