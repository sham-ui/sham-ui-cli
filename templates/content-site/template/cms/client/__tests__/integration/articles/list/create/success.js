import setup, { app } from '../../../helpers';
import axios from 'axios';
jest.mock( 'axios' );

beforeEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'success create article', async() => {
    expect.assertions( 6 );

    const DI = setup();
    axios
        .useDefaultMocks()
        .use( 'get', '/categories', {
            'categories': [
                { 'id': '1', 'name': 'Кухня', 'slug': 'kukhnia' },
                { 'id': '2', 'name': 'Быт', 'slug': 'byt' }
            ]
        } )
        .use( 'get', '/tags', {
            tags: [
                { 'id': '1', 'name': 'Hello world', 'slug': 'hello-world' },
                { 'id': '6', 'name': 'Название тега', 'slug': 'nazvanie-tega' }
            ]
        } )
        .use( 'get', '/articles', {
            'articles': [],
            'meta': {
                'limit': 50, 'offset': 0, 'total': 0
            }
        } )
        .use( 'post', '/articles', {} );

    history.pushState( {}, '', 'http://client.example.com/articles' );
    await app.start( DI );

    app.click( '[data-test-toggle-create-form]' );
    await app.waitRendering();

    expect( window.location.href ).toBe( 'http://client.example.com/articles/new' );

    const formData = {
        title: 'Travel',
        category_id: '2',
        short_body: 'Short body text',
        body: '<p>Full body text</p>',
        tags: [
            { id: '1', name: 'Hello world', slug: 'hello-world' },
            { name: 'New' }
        ],
        published_at: '2000-06-15T01:35:10.000Z'
    };

    app.form.fillBySelector( '[data-test-field-title] input', formData.title );
    app.form.fillBySelector( '[data-test-field-category] select', formData.category_id );
    app.form.fillBySelector( '[data-test-field-short-body] textarea', formData.short_body );
    app.form.fillBySelector( '[data-test-field-body] textarea', formData.body );

    const tagsSelector = '[data-test-field-tags] .autocomplete-input';
    formData.tags.forEach( tag => {
        app.form.fillBySelector( `${tagsSelector} input`, tag.name );
        const item = undefined === tag.slug ?
            'Add as new tag' :
            tag.name
        ;
        app.click( `${tagsSelector} input` );
        app.click(
            `${tagsSelector} .autocomplete-results .result-item[data-test-item="${item}"]`
        );
    } );

    const publishedAtSelector = '[data-test-field-published-at] .sham-ui-datetimepicker';

    app.click( `${publishedAtSelector} th.picker-switch` ); // select month
    app.click( `${publishedAtSelector} th.picker-switch` ); // select year
    app.click( `${publishedAtSelector} th.picker-switch` ); // select year decade
    app.click( `${publishedAtSelector} tbody tr td .decade:nth-child(1)` ); // select first decade
    app.click( `${publishedAtSelector} tbody tr td .year:nth-child(1)` ); // select 2000
    app.click( `${publishedAtSelector} tbody tr td .month:nth-child(6)` ); // select June
    app.click( `${publishedAtSelector} tbody tr:nth-child(3) .day:nth-child(4)` ); // select 15
    app.click( `${publishedAtSelector} tfoot tr th` ); // select time
    app.click( `${publishedAtSelector} tbody tr .hour` ); // select hour
    app.click( `${publishedAtSelector} tbody tr td span:nth-child(10)` ); // select 09
    app.click( `${publishedAtSelector} tbody tr .minute` ); // select minutes
    app.click( `${publishedAtSelector} tbody tr td span:nth-child(8)` ); // select 35
    app.click( `${publishedAtSelector} tbody tr .second` ); // select seconds
    app.click( `${publishedAtSelector} tbody tr td span:nth-child(3)` ); // select 10

    app.checkMainPanel();

    app.form.submit();

    app.click( '[data-test-modal] [data-test-ok-button]' );
    await app.waitRendering();
    app.checkMainPanel();

    expect( axios.mocks.post ).toHaveBeenCalledTimes( 1 );
    expect( axios.mocks.post.mock.calls[ 0 ][ 0 ] ).toBe( '/articles' );
    expect( axios.mocks.post.mock.calls[ 0 ][ 1 ] ).toEqual( formData );
} );

