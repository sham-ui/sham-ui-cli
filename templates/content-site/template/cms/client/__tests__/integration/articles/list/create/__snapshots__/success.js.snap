// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success create article 1`] = `
"<h4 class=\\"page-header\\"> Create article
  <!--0-->
</h4>
<div class=\\"page-body\\">
  <div class=\\"articles-page\\">
    <!--0-->
    <!--0-->
    <div class=\\"panel\\">
      <!--0-->
      <!--0-->
      <form class=\\"form-layout\\">
        <!--0-->
        <!--0-->
        <!--1-->
        <div class=\\"form-group\\" data-test-field-title=\\"\\">
          <label>Title:</label>
          <input autofocus=\\"\\">
        </div>
        <div class=\\"form-group\\" data-test-field-category=\\"\\">
          <label>Category:</label>
          <select>
            <option value=\\"null\\">- - -</option>
            <option value=\\"1\\">Кухня</option>
            <option value=\\"2\\">Быт</option>
          </select>
          <!--0-->
        </div>
        <div class=\\"form-group\\" data-test-field-short-body=\\"\\">
          <label>Short body:</label>
          <textarea></textarea>
        </div>
        <div class=\\"form-group\\" data-test-field-body=\\"\\">
          <label>Body:</label>
          <textarea>&lt;p&gt;Full body text&lt;/p&gt;</textarea>
          <!--1-->
        </div>
        <div class=\\"form-group\\" data-test-field-tags=\\"\\">
          <label>Tags:</label>
          <div class=\\"tags-input\\">
            <div class=\\"tag\\">Hello world<span class=\\"remove\\" data-test-close-button=\\"\\" data-test-remove-tag=\\"Hello world\\">×</span></div>
            <div class=\\"tag\\">New<span class=\\"remove\\" data-test-close-button=\\"\\" data-test-remove-tag=\\"New\\">×</span></div>
            <!--0-->
            <div class=\\"autocomplete-input\\">
              <input autocomplete=\\"off\\">
              <!--0-->
            </div>
            <!--1-->
          </div>
          <!--2-->
        </div>
        <div class=\\"form-group\\" data-test-field-published-at=\\"\\">
          <label>Published at:</label>
          <table class=\\"sham-ui-datetimepicker\\">
            <thead>
              <tr>
                <th colspan=\\"5\\" class=\\"picker-switch\\">четверг, 15 июня 2000 г.</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td class=\\"change\\"><span>+</span></td>
                <td></td>
                <td class=\\"change\\"><span>+</span></td>
                <td></td>
                <td class=\\"change\\"><span>+</span></td>
              </tr>
              <!--0-->
              <tr>
                <td class=\\"hour\\">09</td>
                <td>:</td>
                <td class=\\"minute\\">35</td>
                <td>:</td>
                <td class=\\"second\\">10</td>
              </tr>
              <tr>
                <td class=\\"change\\"><span>–</span></td>
                <td></td>
                <td class=\\"change\\"><span>–</span></td>
                <td></td>
                <td class=\\"change\\"><span>–</span></td>
              </tr>
              <!--1-->
            </tbody>
            <!--0-->
          </table>
          <!--3-->
        </div>
        <!--0-->
        <!--2-->
        <button type=\\"submit\\">
          <!--3-->Save
          <!--4-->
        </button>
        <!--5-->
      </form>
      <!--0-->
      <!--1-->
      <!--1-->
      <!--0-->
    </div>
  </div>
  <!--1-->
</div>
<!--0-->
<!--0-->
<!--0-->"
`;

exports[`success create article 2`] = `
"<h4 class=\\"page-header\\"> Articles
  <!--0-->
</h4>
<div class=\\"page-body\\">
  <div class=\\"articles-page\\">
    <!--0-->
    <!--0-->
    <div class=\\"container mg-b-30\\">
      <a data-test-toggle-create-form=\\"\\" class=\\"btn\\" href=\\"/articles/new\\"> Create new article </a>
    </div>
    <div class=\\"container\\">
      <div class=\\"article-list\\">
        <div class=\\"panel\\">
          <div class=\\"pagination\\">
            <div class=\\"page-buttons\\">
              <!--0-->
              <button data-test-pagination-page=\\"1\\" disabled=\\"\\">1</button>
              <!--1-->
            </div>
            <!--0-->
            <table class=\\"list-table\\">
              <thead>
                <tr>
                  <th>ID</th>
                  <th>Title</th>
                  <th>Slug</th>
                  <th>Category</th>
                  <th>Published at</th>
                  <th></th>
                </tr>
              </thead>
              <tbody>
                <!--0-->
              </tbody>
            </table>
            <!--1-->
            <!--0-->
            <!--0-->
            <!--1-->
            <div class=\\"page-buttons\\">
              <!--0-->
              <button data-test-pagination-page=\\"1\\" disabled=\\"\\">1</button>
              <!--1-->
            </div>
            <!--2-->
          </div>
        </div>
      </div>
    </div>
  </div>
  <!--1-->
</div>
<!--0-->
<!--0-->
<!--0-->"
`;
